### PR TITLE
*: add more tests for metadata lock and fix data race

### LINKED ISSUE #37927: DATA RACE in the metadatalocktest
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
==================
WARNING: DATA RACE
Read at 0x00c006c54f90 by goroutine 10335:
  github.com/pingcap/tidb/sessionctx/variable.(*SessionVars).GetRelatedTableForMDL()
      sessionctx/variable/session.go:3055 +0x116
  github.com/pingcap/tidb/session.RemoveLockDDLJobs()
      session/session.go:3539 +0x93
  github.com/pingcap/tidb/server.(*Server).CheckOldRunningTxn()
      server/server.go:881 +0x1f6
  github.com/pingcap/tidb/domain.(*Domain).mdlCheckLoop()
      domain/domain.go:692 +0x6ea
  github.com/pingcap/tidb/domain.(*Domain).mdlCheckLoop-fm()
      <autogenerated>:1 +0x39
  github.com/pingcap/tidb/util.(*WaitGroupWrapper).Run.func1()
      util/wait_group_wrapper.go:33 +0x73
Previous write at 0x00c006c54f90 by goroutine 81:
  github.com/pingcap/tidb/sessionctx/variable.(*TransactionContext).Cleanup()
      sessionctx/variable/session.go:328 +0x1cb
  github.com/pingcap/tidb/session.(*session).CommitTxn()
      session/session.go:1018 +0x466
  github.com/pingcap/tidb/session.autoCommitAfterStmt()
      session/tidb.go:283 +0x3da
  github.com/pingcap/tidb/session.finishStmt()
      session/tidb.go:249 +0x3ab
  github.com/pingcap/tidb/session.runStmt()
      session/session.go:2215 +0x9ae
  github.com/pingcap/tidb/session.(*session).ExecuteStmt()
      session/session.go:2068 +0x1005
  github.com/pingcap/tidb/testkit.(*TestKit).ExecWithContext()
      testkit/testkit.go:278 +0x5bb
  github.com/pingcap/tidb/testkit.(*TestKit).MustExecWithContext()
      testkit/testkit.go:118 +0xb7
  github.com/pingcap/tidb/testkit.(*TestKit).MustExec()
      testkit/testkit.go:113 +0x95b
  ddl/metadatalocktest/metadatalocktest_test.TestMDLBasicSelect()
      ddl/metadatalocktest/mdl_test.go:63 +0x91f
  testing.tRunner()
      GOROOT/src/testing/testing.go:1446 +0x216
  testing.(*T).Run.func1()
      GOROOT/src/testing/testing.go:1493 +0x47
Goroutine 10335 (running) created at:
  github.com/pingcap/tidb/util.(*WaitGroupWrapper).Run()
      util/wait_group_wrapper.go:31 +0xe4
  github.com/pingcap/tidb/domain.(*Domain).Init()
      domain/domain.go:1026 +0x14fb
  github.com/pingcap/tidb/session.(*domainMap).Get.func1()
      session/tidb.go:93 +0x6c4
  github.com/pingcap/tidb/util.RunWithRetry()
      util/misc.go:69 +0x93
  github.com/pingcap/tidb/session.(*domainMap).Get()
      session/tidb.go:76 +0x258
  github.com/pingcap/tidb/session.createSessionWithOpt()
      session/session.go:2882 +0x75
  github.com/pingcap/tidb/session.createSession()
      session/session.go:2878 +0x84
  github.com/pingcap/tidb/session.createSessions()
      session/session.go:2867 +0x6e
  github.com/pingcap/tidb/session.BootstrapSession()
      session/session.go:2743 +0x337
  github.com/pingcap/tidb/testkit.bootstrap()
      testkit/mockstore.go:72 +0x8c
  github.com/pingcap/tidb/testkit.CreateMockStoreAndDomain()
      testkit/mockstore.go:63 +0xe9
  ddl/metadatalocktest/metadatalocktest_test.TestMDLBasicSelect()
      ddl/metadatalocktest/mdl_test.go:32 +0x58
  testing.tRunner()
      GOROOT/src/testing/testing.go:1446 +0x216
  testing.(*T).Run.func1()
      GOROOT/src/testing/testing.go:1493 +0x47
Goroutine 81 (running) created at:
  testing.(*T).Run()
      GOROOT/src/testing/testing.go:1493 +0x75d
  testing.runTests.func1()
      GOROOT/src/testing/testing.go:1846 +0x99
  testing.tRunner()
      GOROOT/src/testing/testing.go:1446 +0x216
  testing.runTests()
      GOROOT/src/testing/testing.go:1844 +0x7ec
  testing.(*M).Run()
      GOROOT/src/testing/testing.go:1726 +0xa84
  go.uber.org/goleak.VerifyTestMain()
      external/org_uber_go_goleak/testmain.go:53 +0x59
  ddl/metadatalocktest/metadatalocktest_test.TestMain()
      ddl/metadatalocktest/main_test.go:43 +0x338
  main.main()
      bazel-out/k8-fastbuild/bin/ddl/metadatalocktest/metadatalocktest_test_/testmain.go:146 +0x5e8
================== 
```

### PR DESCRIPTION
**Pull request by wjhuang2016**
Signed-off-by: wjhuang2016 <huangwenjun1997@gmail.com>

### What problem does this PR solve?

Issue Number: close #37927

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: ddl/metadatalocktest/mdl_test.go [hunk 1]
```diff
@@ -993,3 +993,115 @@ func TestMDLSavePoint(t *testing.T) {
...
+	tk.MustExec("create table t(a int);")
+	tk.MustExec("insert into t values(1);")
+
+	tk.MustExec("begin")
+	tk.MustQuery("select * from t;")
+	tk.MustGetErrCode("select * from t1;", mysql.ErrNoSuchTable)
+
+	tkDDL.MustExec("create table test.t1(a int);")
+
+	tk.MustGetErrCode("select * from t1;", mysql.ErrNoSuchTable)
```
Comment (Benjamin2037) [inline ddl/metadatalocktest/mdl_test.go:1014]: In RC， the second time， should t1 be seen?

Comment (wjhuang2016) [reply ddl/metadatalocktest/mdl_test.go:1014]: No for now, which can be optimized later.

**Review by hawkingrei — APPROVED** [review]

**Review by Benjamin2037 — APPROVED** [review]
LGTM

**Comment by hawkingrei** [comment]
@tidb-configuration-reviewer PTAL

**Review by bb7133 — APPROVED** [review]
LGTM

**Comment by hawkingrei** [comment]
/hold

Have two failed test cases.

**Comment by hawkingrei** [comment]
@wjhuang2016 


```
        	Error Trace:	/home/jenkins/.tidb/tmp/63a9840cd0739f2c243bb46478607469/sandbox/processwrapper-sandbox/5844/execroot/__main__/bazel-out/k8-fastbuild/bin/planner/cascades/cascades_test_/cascades_test.runfiles/__main__/planner/cascades/result.go:49
        	            				/home/jenkins/.tidb/tmp/63a9840cd0739f2c243bb46478607469/sandbox/processwrapper-sandbox/5844/execroot/__main__/bazel-out/k8-fastbuild/bin/planner/cascades/cascades_test_/cascades_test.runfiles/__main__/planner/cascades/integration_test.go:346
        	Error:      	Not equal: 
        	            	expected: "[1   test Sleep 0 autocommit select * from information_schema.processlist 4b5e7cdd5d3ed84d6c1a6d56403a3d512554b534313caf296268abdec1c9ea99 0 0 ]\n"
        	            	actual  : "[12   test Sleep 0 autocommit select * from information_schema.processlist 4b5e7cdd5d3ed84d6c1a6d56403a3d512554b534313caf296268abdec1c9ea99 0 0 ]\n"
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1,2 +1,2 @@
        	            	-[1   test Sleep 0 autocommit select * from information_schema.processlist 4b5e7cdd5d3ed84d6c1a6d56403a3d512554b534313caf296268abdec1c9ea99 0 0 ]
        	            	+[12   test Sleep 0 autocommit select * from information_schema.processlist 4b5e7cdd5d3ed84d6c1a6d56403a3d512554b534313caf296268abdec1c9ea99 0 0 ]
        	            	 
        	Test:       	TestMemTableScan
        	Messages:   	sql:select * from information_schema.processlist, args:[]
```

**Pull request merged**